tools/dpgen/generate.go

```diff
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package main
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/btf"
+	"github.com/mitchellh/go-wordwrap"
+
+	"github.com/cilium/cilium/pkg/datapath/config"
+
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+)
+
+// Common acronyms to transform into stylized form for Go field names.
+var stylized = map[string]string{
+	"bpf":     "BPF",
+	"lxc":     "LXC",
+	"xdp":     "XDP",
+	"ipv4":    "IPv4",
+	"ipv6":    "IPv6",
+	"nat":     "NAT",
+	"mac":     "MAC",
+	"mtu":     "MTU",
+	"id":      "ID",
+	"ip":      "IP",
+	"netns":   "NetNS",
+	"ipcache": "IPCache",
+}
+
+// varsToStruct generates a Go struct from the configuration variables in the
+// CollectionSpec.
+func varsToStruct(spec *ebpf.CollectionSpec, name, kind, comment string, embeds []string) (string, error) {
+	type field struct {
+		comment  string
+		goName   string
+		cName    string
+		typ      string
+		defValue string
+	}
+
+	kind = "kind:" + kind
+
+	fields := make([]field, 0, len(spec.Variables))
+
+	for n, v := range spec.Variables {
+		// Only consider variables in a specific config section to avoid interfering
+		// with unrelated objects.
+		if v.MapName() != config.Section {
+			continue
+		}
+
+		// DECLARE_CONFIG prefixes the variable name with a well-known prefix to
+		// avoid collisions with other variables with common names.
+		n = strings.TrimPrefix(n, config.ConstantPrefix)
+
+		if v.Type() == nil {
+			return "", fmt.Errorf("variable %s has no type information, was the ELF built without BTF?", n)
+		}
+
+		// Skip variables that don't have the requested kind.
+		tags := v.Type().Tags
+		if !slices.Contains(tags, kind) {
+			continue
+		}
+
+		// Pop the kind tag from the list of tags, we don't want to render it out
+		// to the config struct.
+		tags = slices.DeleteFunc(tags, func(s string) bool { return s == kind })
+
+		if len(tags) == 0 || tags[0] == "" {
+			return "", fmt.Errorf("variable %s has no doc comment", n)
+		}
+
+		typ, err := btfVarGoType(v.Type())
+		if err != nil {
+			return "", fmt.Errorf("variable %s: getting Go type: %w", n, err)
+		}
+
+		comment, err := tagsToComment(tags)
+		if err != nil {
+			return "", fmt.Errorf("variable %s: converting tags to comments: %w", n, err)
+		}
+
+		defValue, err := varGoValue(v)
+		if err != nil {
+			return "", fmt.Errorf("variable %s: getting default Go value: %w", n, err)
+		}
+
+		fields = append(fields, field{comment, camelCase(n), n, typ, fmt.Sprintf("%#v", defValue)})
+	}
+
+	slices.SortStableFunc(fields, func(a, b field) int {
+		return strings.Compare(a.goName, b.goName)
+	})
+
+	var b strings.Builder
+
+	// Render a Go type definition for a configuration struct.
```

Comment:
Sort of a drive-by comment, but what about using `go/ast` and `go/printer` to write these out instead of doing it by hand?

Reply:
I imagine that to be useful when modifying existing sources, parsing into an AST, modifying, then printing back. But I have some doubts about generating the AST for full files. Sounds like a lot of work and I am not sure what the gain would be. (have never done it myself, so just guessing here from other AST experiences)

Reply:
Good observation! The way I've done it here is incredibly common, since Go templates show up in grep and other search tools all the same, and it keeps things easy to modify. I've personally only seen ast being used in linters and code transformers, not so much to generate (due to what Dylan said). A quick template + gofmt gets you to the same spot much quicker.